### PR TITLE
Laboratory work #3, Damir Khamidullin - 22FPL2

### PR DESCRIPTION
**Pull request by Damir-Khamidullin**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -93,6 +119,22 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
         """
+        if not isinstance(text, str):
+            return None
+        tokenized_text = self._tokenize(text)
+        if not tokenized_text:
+            return None
+        for symbol in tokenized_text:
+            self._put(symbol)
+        encoded_corpus = []
+        for symbol in tokenized_text:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:130]: Why do you do this in the second `for`? Let's try to opmtimize and do it one loop

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -159,7 +227,20 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
-
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        decoded_str = ''
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:232]: String concatination is bad idea.

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -388,6 +562,35 @@ def run(self, prompt: str, seq_len: int) -> Optional[str]:
...
         """
+        if (not isinstance(prompt, str) or not prompt or
+            not isinstance(seq_len, int) or seq_len <= 0):
+            return None
+        prompt_encoded = self._text_processor.encode(prompt)
+        if not prompt_encoded:
+            return None
+        sequence_candidates = {prompt_encoded: 0.0}
+        for prediction in range(seq_len):
+            copy = sequence_candidates.copy()
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:573]: do not use built in names for your variables

File: lab_3_generate_by_ngrams/start.py [hunk 4]
```diff
@@ -11,9 +13,24 @@ def main() -> None:
...
+    greedytextgenerator = GreedyTextGenerator(ngramlanguagemodel_2, textprocessor)
+    predicted_text = greedytextgenerator.run(51, 'Vernon')
+    print(predicted_text)
+    beamsearchtextgenerator = BeamSearchTextGenerator(ngramlanguagemodel_2, textprocessor, 7)
+    predicted_text_2 = beamsearchtextgenerator.run('Vernon', 56)
+    print(predicted_text_2)
     assert result
 
 
+
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/start.py:34]: I do not think you really need this blank line

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/start.py (hunk 4 above)
Comment (Damir-Khamidullin) [reply lab_3_generate_by_ngrams/start.py]: deleted

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (Damir-Khamidullin) [reply lab_3_generate_by_ngrams/main.py]: renamed

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (Damir-Khamidullin) [reply lab_3_generate_by_ngrams/main.py]: changed

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (Damir-Khamidullin) [reply lab_3_generate_by_ngrams/main.py:128]: merged

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -104,6 +145,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:150]: reverse the condition

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -129,6 +184,13 @@ def fill_from_ngrams(self, content: dict) -> None:
         Args:
             content (dict): ngrams from external JSON
         """
+        if not isinstance(content, dict) or not content:
+            return None
+        for n_gram in content['freq']:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:189]: for in for... Do you know better solution?

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -159,7 +232,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
-
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        decoded_list = list(decoded_corpus)
+        if decoded_list[-1] == self._end_of_word_token:
+            decoded_list = decoded_list[:-1]
+        decoded_str = ''.join(decoded_list).replace(self._end_of_word_token, ' ') + '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:240]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -208,6 +295,15 @@ def build(self) -> int:
         In case of corrupt input arguments or methods used return None,
         1 is returned
         """
+        if not isinstance(self._encoded_corpus, tuple) or not self._encoded_corpus:
+            return 1
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not n_grams:
+            return 1
+        for n_gram in set(n_grams):
+            b_freq = len([n_gram_2 for n_gram_2 in n_grams if n_gram_2[:-1] == n_gram[:-1]])
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:304]: count

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -270,6 +383,26 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+        n_gram_size = self._model.get_n_gram_size()
+        if not prompt_encoded or not n_gram_size:
+            return None
+        list_encoded = list(prompt_encoded)
+        for prediction in range(seq_len):
+            seq = tuple(list_encoded[-n_gram_size + 1:])
+            variants = self._model.generate_next_token(seq)
+            if not variants:
+                break
+            best_choice = sorted([pred for pred, freq in variants.items() if
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:399]: filter

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:150]: !

File: lab_3_generate_by_ngrams/main.py (hunk 9 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:399]: and too hard to understand

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**